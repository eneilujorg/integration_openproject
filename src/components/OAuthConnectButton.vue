<template>
	<button v-if="!!requestUrl"
		class="oauth-connect--button"
		@click="onOAuthClick">
		<span class="icon icon-external" />
		{{ t('integration_openproject', 'Connect to OpenProject') }}
	</button>
	<div v-else class="oauth-connect--message">
		{{ adminConfigNotOkMessage }}
	</div>
</template>
<script>
import axios from '@nextcloud/axios'
import { generateUrl } from '@nextcloud/router'
import { showError } from '@nextcloud/dialogs'
import { translate as t } from '@nextcloud/l10n'

export default {
	name: 'OAuthConnectButton',

	props: {
		requestUrl: {
			type: [String, Boolean],
			required: true,
		},
	},

	computed: {
		adminConfigNotOkMessage() {
			return t('integration_openproject', 'Some OpenProject integration application settings are not working.'
				+ ' Please contact your Nextcloud administrator.')
		},
	},

	methods: {
		generateRandomString(length) {
			let text = ''
			const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~'
			for (let i = 0; i < length; i++) {
				text += possible.charAt(Math.floor(Math.random() * possible.length))
			}
			return text
		},
		async digest(text) {
			return await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text))
		},
		async generateCodeChallenge(codeVerifier) {
			const digest = await this.digest(codeVerifier)
			return btoa(String.fromCharCode(...new Uint8Array(digest)))
				.replace(/=/g, '')
				.replace(/\+/g, '-')
				.replace(/\//g, '_')
		},
		async onOAuthClick() {
			const oauthState = this.generateRandomString(10)
			const codeVerifier = this.generateRandomString(128)
			const codeChallenge = await this.generateCodeChallenge(codeVerifier)
			const requestUrl = this.requestUrl
				+ '&state=' + encodeURIComponent(oauthState)
				+ '&code_challenge=' + codeChallenge
				+ '&code_challenge_method=S256'

			const req = {
				values: {
					oauth_state: oauthState,
					code_verifier: codeVerifier,
				},
			}
			const url = generateUrl('/apps/integration_openproject/config')
			axios.put(url, req)
				.then(() => {
					window.location.replace(requestUrl)
				})
				.catch((error) => {
					showError(
						t('integration_openproject', 'Failed to save OpenProject OAuth state')
						+ ': ' + error.message
					)
				})
		},
	},
}
</script>
<style lang="scss" scoped>
.oauth-connect {
	&--message {
		font-size: 1rem;
		text-align: center;
		font-weight: 400;
		color: #878787;
		padding: 0px 18px;
		line-height: 1.4rem;
	}
}
</style>
