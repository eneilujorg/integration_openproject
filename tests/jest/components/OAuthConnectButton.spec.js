/* jshint esversion: 8 */

import { shallowMount, createLocalVue } from '@vue/test-utils'
import OAuthConnectButton from '../../../src/components/OAuthConnectButton.vue'
import axios from '@nextcloud/axios'
import * as dialogs from '@nextcloud/dialogs'

jest.mock('@nextcloud/axios')
jest.mock('@nextcloud/dialogs')

const { location } = window
const localVue = createLocalVue()

describe('OAuthConnectButton.vue Test', () => {
	let wrapper
	beforeEach(() => {
		delete window.location
		window.location = { replace: jest.fn() }
		wrapper = shallowMount(OAuthConnectButton, {
			localVue,
			mocks: {
				t: (app, msg) => msg,
			},
			propsData: {
				requestUrl: 'http://openproject/oauth/',
			},
		})
	})
	afterEach(() => {
		window.location = location
	})
	describe('on successful saving of the state', () => {
		beforeEach(() => {
			axios.put.mockImplementationOnce(() =>
				Promise.resolve({}),
			)
		})
		it('saves the state to user config', async () => {
			wrapper.find('button').trigger('click')
			await localVue.nextTick()
			expect(axios.put).toHaveBeenCalledWith(
				'http://localhost/apps/integration_openproject/config',
				{ values: { oauth_state: expect.stringMatching(/[a-z0-9]{9}/) } },
			)
		})
		it('redirects to the openproject oauth uri', async () => {
			wrapper.find('button').trigger('click')
			await localVue.nextTick()
			expect(window.location.replace).toHaveBeenCalledWith(
				expect.stringMatching(/http:\/\/openproject\/oauth\/&state=[a-z0-9]{9}/),
			)
		})
	})
	describe('on unsuccessful saving of the state', () => {
		beforeEach(() => {
			const err = new Error()
			err.message = 'some issue'
			axios.put.mockRejectedValueOnce(err)
		})
		it('shows an error', async () => {
			dialogs.showError.mockImplementationOnce()
			wrapper.find('button').trigger('click')
			await localVue.nextTick()
			expect(dialogs.showError).toHaveBeenCalledWith(
				'Failed to save OpenProject OAuth state: some issue'
			)
			expect(window.location.replace).not.toHaveBeenCalled()
		})
	})
})
