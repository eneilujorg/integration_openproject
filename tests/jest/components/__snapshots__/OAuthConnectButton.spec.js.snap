// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuthConnectButton.vue Test when the admin config is not okay should show message 1`] = `
<div class="oauth-connect--message">
  Some OpenProject integration application settings are not working. Please contact your Nextcloud administrator.
</div>
`;
